app/assets/javascripts/govuk_publishing_components/components/autocomplete.js

```diff
@@ -0,0 +1,561 @@
+          if (!immediate) func.apply(context, args)
+        }
+
+        var callNow = immediate && !timeout
+        clearTimeout(timeout)
+        timeout = setTimeout(later, wait)
+
+        if (callNow) func.apply(context, args)
+      }
+    },
+    getAriaLabel: function (labelStr) {
+      if (labelStr?.length) {
+        var isLabelId = labelStr.startsWith('#')
+
+        return {
+          attribute: isLabelId ? 'aria-labelledby' : 'aria-label',
+          content: isLabelId ? labelStr.substring(1) : labelStr
+        }
+      }
+    },
+    isPromise: function (value) {
+      return Boolean(value && typeof value.then === 'function')
+    },
+    isFunction: function (value) {
+      return Boolean(typeof value === 'function')
+    },
+    isArray: function (string) {
+      try {
+        return Array.isArray(JSON.parse(string))
+      } catch (e) {
+        return false
+      }
+    }
+  }
+
+  // Creates a props object with overridden toString function. toString returns an attributes
+  // string in the format: `key1='value1' key2='value2'` for easy use in an HTML string.
+  function Props (index, selectedIndex, baseClass) {
+    this.id = `${baseClass}-result-${index}`
+    this.class = `${baseClass}__result`
+    this['data-result-index'] = index
+    this.role = 'option'
+    if (index === selectedIndex) {
+      this['aria-selected'] = 'true'
+    }
+  }
+
+  Props.prototype.toString = function () {
+    return Object.keys(this).reduce(
+      (str, key) => `${str} ${key}='${this[key]}'`,
+      ''
+    )
+  }
+
+  function AutocompleteCore (
+    search,
+    autoSelect,
+    setValue,
+    setAttribute,
+    onUpdate,
+    onSubmit,
+    onShow,
+    onHide,
+    onLoading,
+    onLoaded,
+    submitOnSelect
+  ) {
+    this.value = ''
+    this.searchCounter = 0
+    this.results = []
+    this.selectedIndex = -1
+    this.selectedResult = null
+
+    switch (true) {
+      case util.isFunction(search):
+        this.search = (value) => Promise.resolve(search(value))
+        break
+      case util.isArray(search):
+        this.search = () => Promise.resolve(JSON.parse(search))
+        break
+      default:
+        this.search = async (value) => {
+          var url = new URL(search)
+          url.searchParams.append('q', value)
+
+          var response = await fetch(url)
+          var completes = await response.json()
+          return completes
+        }
+        break
+    }
+
+    this.autoSelect = autoSelect || false
+    this.setValue = setValue || function () { }
+    this.setAttribute = setAttribute || function () { }
+    this.onUpdate = onUpdate || function () { }
+    this.onSubmit = onSubmit || function () { }
+    this.onShow = onShow
+    this.onHide = onHide || function () { }
+    this.onLoading = onLoading || function () { }
+    this.onLoaded = onLoaded || function () { }
+    this.submitOnSelect = submitOnSelect || false
+  }
+
+  AutocompleteCore.prototype.destroy = function () {
```

Comment:
Why are there two `destroy` functions - is it not possible to combine them into one? And when is this called - as far as I can see the autocomplete doesn't get destroyed during normal user operation.